### DISCUSSION TITLE
Nodes running in high availability (with k8s) have duplicated apps/dbs

### DISCUSSION THREAD

**Discussion by jaysonsantos** [gravitational/teleport#7389] [Q&A]
Hey there folks, I am not opening an issue because maybe there is a solution for this.
I am using the example chart for the kube agent with multiple replicas, as they read the same config, they all report to the proxy (and then auth, as my k8s is behind a nat) multiple instances of the same application and/or database.
Also, when restarting the deployment, new pods come up until the old ones die and at this time, the number of duplication double until (assuming) the state maintainer process collects the garbage.
To solve that I thought of using instead of emptyDir some persistent volume but that would limit the deployment replica to 1, unless I use ReadWriteMany (which could potentially lead to race conditions as both processes would write on the same place).
Do you have any other ideas on how to avoid the duplication for the same name and uri?
Have you also used teleport nodes with k8s, what can you say about your experience with it?
Thanks in advance,
Jayson

teleport-kube-agent with version 6.2.3
teleport-cluster (proxy + auth) in another k8s cluster with version 6.2.3

**Comment by webvictim** [comment]
This is a known issue for applications and is being tracked in https://github.com/gravitational/teleport/issues/5580. This was fixed recently for database services (the fix will be in Teleport 7.0) but has not yet been addressed for applications.

The only workaround I can think of is limiting the number of agents to 1 and using some persistent storage to make sure that the agent doesn't re-register itself as a new service every time it restarts. We have a PR (https://github.com/gravitational/teleport/pull/7123) in review for supporting persistent storage in the `teleport-kube-agent` chart which should make this a little easier.

**Reply by jaysonsantos** [reply]
Hey there @webvictim does this mean that in version 7 teleport would be able to balance connections to databases?
I did exactly what you said, I've used EFS for a persisent storage, set replicas as 1 and it does the but I have an automated teleport configuration that fills the databases and it has to restart the pod making a small downtime.
I saw that there is the USRx signal but It got me afraid that it would pileup until the point that the memory is not able to handle too many forked processes. Would be nice if there was a signal to reload the configurations, should I open an issue or do you recall one about that?
Thanks for the answers!

**Reply by webvictim** [reply]
Yes, Teleport should be able to load balance connections to databases in v7.

Could you maybe write the config via `initContainer` before the pod starts rather than needing to restart?

`kill -HUP` should work to reload the config and fork but like you say, a restart occasionally would probably help. There isn't a straight "reload the config in the current process" command.

**Reply by jaysonsantos** [reply]
I sort of did what the initContainer would do but I've preferred to split the responsibility there and I have another service that listens for changes in RDS or applications within the cluster to generate the configmap.
And I also tried SIGHUPping the process but as the main process dies, kubernetes thinks that is has exited. The only way I could think of that working is with by the forked process replacing the parent with something like exec/disown and on v6 this is not there, right?

**Reply by webvictim** [reply]
Yeah, you'd likely need some kind of wrapper around the Teleport process which stays running to keep k8s happy. This is mostly why I suggested that an `initContainer` would be preferable as the config could be generated before the pod starts, then it doesn't need to be restarted in-place. If there are config changes, you'd then relaunch a new pod (to pick up/generate the new config) and let the old one die, etc.

I agree that this should be better, unfortunately we don't really have any time to investigate further or work on it. If you have any ways that we could improve the experience and want to submit a PR, we'll certainly take a look :)

**Reply by jaysonsantos** [reply]
>  If there are config changes, you'd then relaunch a new pod (to pick up/generate the new config) and let the old one die, etc.
sorry, that is what i meant by restart, it actually starts a new deployment rollout making sure the new one is stable.

but i guess that with the version 7 i would be able to scale horizontally to mitigate a bit the new pods creation when there are changes and users connected to databases for example.

thanks for the enlightenment!
